metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java

```diff
@@ -96,49 +110,73 @@
         this.alpha = alpha;
         this.size = size;
         this.count = new AtomicLong(0);
-        reservoirs.add(this);
+        RESERVOIRS.add(this);
         updateCurrentTimeInSeconds();
         this.startTime = currentTimeInSeconds;
         this.nextScaleTime = new AtomicLong(clock.nanoTick() + RESCALE_THRESHOLD);
     }
 
     static void init() {
-        if (EXECUTOR_UPDATER_CONFIGURED.compareAndSet(false, true)) {
+        CURRENT_TIME_EXECUTOR_SERVICE_STATE.getAndUpdate(ExponentiallyDecayingReservoir::startCurrentTimeExecutor);
```

Comment:
Update function can be executed multiple times in case of race against the contention so it needs to be always reentrant.  
I would rather go with states being used as gates:

```java
CompletableFuture<ScheduledExecutorService> currentTimeUpdaterExecutorService = new CompletableFuture();

static void init() {
    if(CURRENT_TIME_EXECUTOR_SERVICE_STATE.compareAndSet(DORMANT, STARTED)){
    // since now exactly one shutdown can happen
   // and it will wait till executor is prepared, bcs exec is a future
   var exec = Executors.newSingleThreadScheduledExecutor();
  
  .....

  // hold possible shutdown until init is done
  currentTimeUpdaterExecutorService.complete(exec);
  }
}

void onServerShutdown() {
    // init can happen here
    if(CURRENT_TIME_EXECUTOR_SERVICE_STATE.compareAndSet(STARTED, STOPPED)){
    // init can't start here
    // but can be still starting(NPE on exec.shutdown()), wait for it to finish
   currentTimeUpdaterExecutorService.get(5, SECONDS).shutdownNow();
....
```

Reply:
Yeah, I went back and forth on this choice but I do see your point. The whole point of introducing the three-state model is to make sure an init cannot run after _or during_ a shutdown. I'll change it.

Reply:
I'm wondering whether it's as effective and perhaps simpler to read and understand if `init` and `onServerShutdown` are `synchronized` instead. That also helps us in the case where the shutdown occurs entirely before `init` so the state is `DORMANT` at shutdown time--unlikely but possible--and we need to mark the state as `STOPPED` in that case to prevent a subsequent `init` from running.

Reply:
Sure you can use locks,  but the state issue will prevail anyway.

I see, DORMANT after shutdown gate can be solved easily:
```java
        void onServerShutdown() {
            if(CURRENT_TIME_EXECUTOR_SERVICE_STATE.getAndSet(STOPPED) == STARTED){
```